### PR TITLE
Blocking command support (BLPOP/BRPOP & BLMOVE)

### PR DESCRIPTION
**Pull request by TalZaccai**
- Adding support for blocking commands in Garnet (for List and SortedSet objects)
- Implementing logic for BRPOP/BLPOP & BLMOVE commands

**Still remaining to address in this PR:**
- [x] Block processing of further commands once a blocking command has been called
- [x] Stop the main loop in CollectionItemBroker if there are no active observers 
- [x] Periodically clean up "zombie" observers (observers whose resp session has disposed but are still on the key's observer queue)
- [x] Remove keys whose observer queues are empty
- [x] Add more test cases

Added some website docs for the ItemBroker class here: _**website/docs/dev/collection-broker.md**_

### CODE REVIEW

File: libs/server/Resp/Objects/ListCommands.cs [hunk 1]
```diff
@@ -191,6 +192,60 @@ private unsafe bool ListPop<TGarnetApi>(int count, byte* ptr, ListOperation lop,
...
+                    return false;
+                return true;
+            }
+
+            if (!RespReadUtils.ReadDoubleWithLengthHeader(out var timeout, out var parsed, ref ptr,
+                recvBufferPtr + bytesRead) || !parsed)
+                return false;
+
+            var arrKeys = keys.Select(k => k.Span.ToArray()).ToArray();
+            var result = itemBroker.GetListItemAsync(arrKeys, lop, this, timeout).Result;
```
Comment (badrishc) [inline libs/server/Resp/Objects/ListCommands.cs:225]: Wait on a SemaphoreSlim here, that can be canceled (using a cancellation token) if the socket is killed or the session dies for some reason (i.e., session Dispose gets called).

**Comment by badrishc** [comment]
Please report BDN results of main vs. this branch for `-f *.RespParseStress.*`

File: libs/server/Objects/ItemBroker/CollectionItemBroker.cs [hunk 2]
```diff
@@ -0,0 +1,578 @@
...
+    /// with the desired object type and operation and a list of keys to the desired objects.
+    /// When an item is added to a collection, the StorageSession will call the Publish method with the relevant object key
+    /// to notify the broker that a new item may be available.
+    /// The main loop, in the Start method, listens for published item additions as well as new observers
+    /// and notifies the calling method if an item was found.
+    /// </summary>
+    public class CollectionItemBroker : IDisposable
+    {
+        // Queue of events to be handled by the main loops
+        private readonly AsyncQueue<BrokerEventBase> brokerEventsQueue = new();
```
Comment (badrishc) [inline libs/server/Objects/ItemBroker/CollectionItemBroker.cs:27]: can these heavy components be instantiated lazily? i.e. if user is not using blocking collections, we should not incur either allocations or tasks.

Comment (badrishc) [reply libs/server/Objects/ItemBroker/CollectionItemBroker.cs:27]: same holds for all the other allocations below.

File: libs/server/Objects/ItemBroker/CollectionItemBroker.cs [hunk 3]
```diff
@@ -0,0 +1,578 @@
...
+        /// <param name="srcKey">Key of object to observe</param>
+        /// <param name="dstKey">Key of object to add item into</param>
+        /// <param name="session">Calling session instance</param>
+        /// <param name="timeoutInSeconds">Timeout of operation (in seconds, 0 for waiting indefinitely)</param>
+        /// <param name="cmdArgs">Additional command arguments</param>
+        /// <returns>Result of operation</returns>
+        internal async Task<CollectionItemResult> MoveCollectionItemAsync(RespCommand command, byte[] srcKey, byte[] dstKey
+            , RespServerSession session, double timeoutInSeconds, object[] cmdArgs = null)
+        {
+            var args = cmdArgs == null ? new[] { dstKey } : new object[] { dstKey }.Union(cmdArgs).ToArray();
```
Comment (badrishc) [inline libs/server/Objects/ItemBroker/CollectionItemBroker.cs:132]: too many allocations (new) - can this be optimized to have no allocations. we will eventually move to Span<byte> for keys, until then we cannot avoid the byte[] of key, but the other allocations can be avoided.

File: libs/server/Objects/ItemBroker/CollectionItemBroker.cs [hunk 4]
```diff
@@ -0,0 +1,578 @@
...
+                {
+                    // If the key already has a non-empty observer queue, it does not have an item to retrieve
+                    // Otherwise, try to retrieve next available item
+                    if ((keysToObservers.ContainsKey(key) && keysToObservers[key].Count > 0) ||
+                        !TryGetNextItem(key, observer.Session.storageSession, observer.Command, observer.CommandArgs,
+                            out _, out var nextItem)) continue;
+
+                    // An item was found - set the observer result and return
+                    sessionIdToObserver.TryRemove(observer.Session.ObjectStoreSessionID, out _);
+                    observer.HandleSetResult(new CollectionItemResult(key, nextItem));
```
Comment (badrishc) [inline libs/server/Objects/ItemBroker/CollectionItemBroker.cs:224]: Is `CollectionItemResult` a class, if yes, this is yet another allocation.

File: libs/server/Objects/ItemBroker/CollectionItemBroker.cs [hunk 5]
```diff
@@ -0,0 +1,578 @@
...
+        /// Try to get next available item from object
+        /// </summary>
+        /// <param name="key">Key of object</param>
+        /// <param name="storageSession">Current storage session</param>
+        /// <param name="command">RESP command</param>
+        /// <param name="cmdArgs">Additional command arguments</param>
+        /// <param name="currCount">Collection size</param>
+        /// <param name="nextItem">Retrieved item</param>
+        /// <returns>True if found available item</returns>
+        private bool TryGetNextItem(byte[] key, StorageSession storageSession, RespCommand command, object[] cmdArgs, out int currCount, out byte[] nextItem)
```
Comment (badrishc) [inline libs/server/Objects/ItemBroker/CollectionItemBroker.cs:410]: why does cmdArgs need to be an object[], can be it a byte[] instead?

File: libs/server/Objects/ItemBroker/CollectionItemBroker.cs [hunk 6]
```diff
@@ -0,0 +1,578 @@
...
+            {
+                // Repeat while not disposed or cancelled
+                while (!disposed && !cts.IsCancellationRequested)
+                {
+                    // Check if current task is done
+                    if (handleNextEvent == null || handleNextEvent.IsCompleted)
+                    {
+                        // Set task to asynchronously dequeue next event in broker's queue
+                        // once event is dequeued successfully, call handler method
+                        handleNextEvent = brokerEventsQueue.DequeueAsync(cts.Token).ContinueWith(t =>
```
Comment (badrishc) [inline libs/server/Objects/ItemBroker/CollectionItemBroker.cs:532]: will there be a task allocation for every item in the queue? for heavy load, can we synchronously dequeue, and only when that fails we actually incur the cost of async?

File: libs/server/Storage/Session/ObjectStore/ListOps.cs [hunk 7]
```diff
@@ -84,6 +85,7 @@ public unsafe GarnetStatus ListPush<TObjectContext>(ArgSlice key, ArgSlice eleme
             var status = RMWObjectStoreOperation(key.ToArray(), element, out var output, ref objectStoreContext);
             itemsDoneCount = output.countDone;
 
+            itemBroker.HandleCollectionUpdate(key.Span.ToArray());
```
Comment (badrishc) [inline libs/server/Storage/Session/ObjectStore/ListOps.cs:88]: ideally the broker deals with spans directly so we are not allocting unless it is absolutely necessary.

**Review by badrishc — CHANGES_REQUESTED** [review]
Added comments mostly related to allocations (both byte[] and Task) and potentially heavy code paths. We need to optimize for:
1. The case where BL* instructions are not used (i.e., heavy classes are created lazily)
2. Optional - avoid byte[] allocations by keeping items in FasterLog
3. Avoid Task allocations by using the async codepath only when no items available to process synchronously.

File: libs/server/Objects/ItemBroker/CollectionItemBroker.cs (hunk 4 above)
Comment (TalZaccai) [reply libs/server/Objects/ItemBroker/CollectionItemBroker.cs:235]: Changed to struct

**Comment by TalZaccai** [comment]
> Please report BDN results of main vs. this branch for `-f *.RespParseStress.*`

**current branch:**
| Method     | Job    | EnvironmentVariables | Runtime  | Mean       | Error     | StdDev    | Allocated |
|----------- |------- |--------------------- |--------- |-----------:|----------:|----------:|----------:|
| InlinePing | .NET 6 | Empty                | .NET 6.0 |   3.736 us | 0.0444 us | 0.0394 us |         - |
| Set        | .NET 6 | Empty                | .NET 6.0 |  40.151 us | 0.5339 us | 0.4994 us |         - |
| SetEx      | .NET 6 | Empty                | .NET 6.0 |  49.062 us | 0.9620 us | 0.8033 us |         - |
| Get        | .NET 6 | Empty                | .NET 6.0 |  23.247 us | 0.4078 us | 0.3815 us |         - |
| ZAddRem    | .NET 6 | Empty                | .NET 6.0 | 203.544 us | 3.8383 us | 3.5903 us |   23552 B |
| InlinePing | .NET 8 | DOTNET_TieredPGO=0   | .NET 8.0 |   3.246 us | 0.0126 us | 0.0098 us |         - |
| Set        | .NET 8 | DOTNET_TieredPGO=0   | .NET 8.0 |  27.550 us | 0.0505 us | 0.0394 us |         - |
| SetEx      | .NET 8 | DOTNET_TieredPGO=0   | .NET 8.0 |  38.817 us | 0.4597 us | 0.4300 us |         - |
| Get        | .NET 8 | DOTNET_TieredPGO=0   | .NET 8.0 |  17.808 us | 0.1461 us | 0.1296 us |         - |
| ZAddRem    | .NET 8 | DOTNET_TieredPGO=0   | .NET 8.0 | 167.105 us | 3.1375 us | 2.7813 us |   23552 B |

**main:**
| Method     | Job    | EnvironmentVariables | Runtime  | Mean       | Error     | StdDev    | Allocated |
|----------- |------- |--------------------- |--------- |-----------:|----------:|----------:|----------:|
| InlinePing | .NET 6 | Empty                | .NET 6.0 |   3.461 us | 0.0321 us | 0.0268 us |         - |
| Set        | .NET 6 | Empty                | .NET 6.0 |  38.975 us | 0.7435 us | 0.7302 us |         - |
| SetEx      | .NET 6 | Empty                | .NET 6.0 |  50.464 us | 0.3800 us | 0.3173 us |         - |
| Get        | .NET 6 | Empty                | .NET 6.0 |  22.726 us | 0.1165 us | 0.0910 us |         - |
| ZAddRem    | .NET 6 | Empty                | .NET 6.0 | 195.731 us | 1.7133 us | 1.4307 us |   23552 B |
| InlinePing | .NET 8 | DOTNET_TieredPGO=0   | .NET 8.0 |   3.376 us | 0.0247 us | 0.0219 us |         - |
| Set        | .NET 8 | DOTNET_TieredPGO=0   | .NET 8.0 |  27.949 us | 0.1557 us | 0.1300 us |         - |
| SetEx      | .NET 8 | DOTNET_TieredPGO=0   | .NET 8.0 |  38.309 us | 0.2966 us | 0.2774 us |         - |
| Get        | .NET 8 | DOTNET_TieredPGO=0   | .NET 8.0 |  18.324 us | 0.1501 us | 0.1404 us |         - |
| ZAddRem    | .NET 8 | DOTNET_TieredPGO=0   | .NET 8.0 | 165.709 us | 3.2412 us | 4.9496 us |   23552 B |

**Review by badrishc — APPROVED** [review]

**Pull request merged**